gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java

```diff
@@ -857,6 +914,13 @@ protected void afterExecute(Runnable r, Throwable t) {
 
           // Compaction job of a dataset is successful.
           jobRunner.getDataset().setState(COMPACTION_COMPLETE);
+
+          // Add a new MPCompactor job if recompaction is necessary.
+          if (jobRunner.getDataset().needToRecompact()) {
```

Comment:
the dataset state should be set to `COMPACTION_COMPLETE` only if it doesn't need to be recompacted. Otherwise `submitCompactionJobsAndWaitForCompletion()` will terminate the job.